File: sw/src/wib_client.cxx
#include <unistd.h>
#include <fstream>
#include <streambuf>

#include <zmq.hpp>
#include <readline/readline.h>
#include <readline/history.h>

#include "wib.pb.h"
#include "femb_3asic.h" //for fast command defines
#include "wib.h" //for spy buffer defines

using namespace std;

void print_usage(const char *prog) {
    glog.log( "Usage: %s [-w ip] [cmd] \n", prog);
}

void print_help() {
    glog.log("Available commands:\n");
    glog.log("  reboot\n");
    glog.log("    Reboot the WIB\n");
    glog.log("  fw_timestamp\n");
    glog.log("    Return the firmware version timestamp\n");
    glog.log("  sw_version\n");
    glog.log("    Return the software build version\n");
    glog.log("  log [boot|clear]\n");
    glog.log("    Return the wib_server log (or return boot log, or clear the logs)\n");
    glog.log("  script filename\n");
    glog.log("    Run a WIB script (local file will be sent, otherwise filename is remote in /etc/wib/)\n");
    glog.log("  daqspy filename\n");
    glog.log("    Read 1MB from each daq spy buffer and write the 2MB data to filename\n");
    glog.log("  peek addr\n");
    glog.log("    Read a 32bit value from WIB address space\n");
    glog.log("  poke addr value\n");
    glog.log("    Write a 32bit value to WIB address space\n");
    glog.log("  cdpeek femb_idx cd_idx chip_addr reg_page reg_addr\n");
    glog.log("    Read a 8bit value from COLDATA I2C address space\n");
    glog.log("  cdpoke femb_idx cd_idx chip_addr reg_page reg_addr data\n");
    glog.log("    Write a 8bit value to COLDATA I2C address space\n");
    glog.log("  cdfastcmd cmd\n");
    glog.log("    Send the fast command cmd to all coldata chips\n");
    glog.log("  update root_archive boot_archive\n");
    glog.log("    Deploy a new root and boot archive to the WIB\n");
    glog.log("  exit\n");
    glog.log("    Closes the command interface\n");
}

template <class R, class C>
void send_command(zmq::socket_t &socket, const C &msg, R &repl) {

    wib::Command command;
    command.mutable_cmd()->PackFrom(msg);
    
    string cmd_str;
    command.SerializeToString(&cmd_str);
    
    zmq::message_t request(cmd_str.size());
    memcpy((void*)request.data(), cmd_str.c_str(), cmd_str.size());
    socket.send(request,zmq::send_flags::none);
    
    zmq::message_t reply;
    socket.recv(reply,zmq::recv_flags::none);
    
    string reply_str(static_cast<char*>(reply.data()), reply.size());
    repl.ParseFromString(reply_str);
    
}

int run_command(zmq::socket_t &s, int argc, char **argv) {
    if (argc < 1) return 1;
    
    string cmd(argv[0]);
    if (cmd == "exit") { 
        return 255;
    } else if (cmd == "get_sensors") {
        wib::GetSensors req;
        wib::GetSensors::Sensors rep;
        send_command(s,req,rep);
    } else if (cmd == "fw_timestamp") {
        wib::GetTimestamp req;
        wib::GetTimestamp::Timestamp rep;
        send_command(s,req,rep);
        glog.log("fw_timestamp code: 0x%08X\n",rep.timestamp());
        glog.log("decoded: %i/%i/%i %i:%i:%i\n",rep.year(),rep.month(),rep.day(),rep.hour(),rep.min(),rep.sec());
    } else if (cmd == "sw_version") {
        wib::GetSWVersion req;
        wib::GetSWVersion::Version rep;
        send_command(s,req,rep);
        glog.log("sw_version: %s\n",rep.version().c_str());
    } else if (cmd == "log") {
        if (argc > 2) {
            glog.log("Usage: log [clear|boot]\n");
            return 0;
        }
        bool boot = false;
        bool clear = false;
        if (argc == 2) {
            string subcmd(argv[1]);
            if (subcmd == "boot") {
                boot = true;
            } else if (subcmd == "clear") {
                clear = true;
            } else {    
                glog.log("Usage: log [clear|boot]\n");
                return 0;
            }
        }
        wib::LogControl req;
        req.set_return_log(!boot && !clear);
        req.set_boot_log(boot);
        req.set_clear_log(clear);
        wib::LogControl::Log rep;
        send_command(s,req,rep);
        const string &log = rep.contents();
        printf("%s",log.c_str());
    } else if (cmd == "script") {
        if (argc != 2) {
            glog.log("Usage: script filename\n");
            return 0;
        }
        string fname(argv[1]);
        ifstream fin(fname);
        if (fin.is_open()) {
            glog.log("Executing local script... ");
            fflush(stdout);
            string script((istreambuf_iterator<char>(fin)), istreambuf_iterator<char>());
            wib::Script req;
            req.set_script(script);
            req.set_file(false);
            wib::Status rep;
            send_command(s,req,rep);
            if (rep.success()) {
                glog.log("Success\n");
            } else {
                glog.log("Failure\n");
            }
            fin.close();
        } else {
            glog.log("Executing remote script... ");
            fflush(stdout);
            wib::Script req;
            req.set_script(fname);
            req.set_file(true);
            wib::Status rep;
            send_command(s,req,rep);
            if (rep.success()) {
                glog.log("Success\n");
            } else {
                glog.log("Failure\n");
            }
        }
    } else if (cmd == "daqspy") {
        if (argc != 2) {
            glog.log("Usage: daqspy filename\n");
            return 0;
        }
        wib::ReadDaqSpy req;
        req.set_buf0(true);
        req.set_buf1(true);
        wib::ReadDaqSpy::DaqSpy rep;
        glog.log("Acquiring DAQ spy buffer...");
        fflush(stdout);
        send_command(s,req,rep);
        string fname(argv[1]);
        ofstream fout(fname,ofstream::binary);
        fout.write(rep.buf0().c_str(),rep.buf0().size());
        fout.write(rep.buf1().c_str(),rep.buf1().size());
        fout.close();
        if (rep.success()) {
            glog.log("Success\n");
        } else {
            glog.log("Failure\n");
        }
    } else if (cmd == "calibrate") {
        if (argc != 1) {
            glog.log("Usage: calibrate\n");
            return 0;
        }
        wib::Calibrate req;
        wib::Status rep;
        send_command(s,req,rep);
        if (rep.success()) {
            glog.log("Success\n");
        } else {
            glog.log("Failure\n");
        }
    } else if (cmd == "update") {
        if (argc != 3) {
            glog.log("Usage: update root_archive boot_archive\n");
            return 0;
        }
        
        size_t length;
        string root_archive, boot_archive;
        
        ifstream in_root(argv[1], ios::binary);
        if (!in_root.is_open()) {
            glog.log("Could not open root archive: %s\n",argv[1]);
            return 0;
        }
        in_root.ignore( numeric_limits<streamsize>::max() );
        length = in_root.gcount();
        in_root.clear();
        in_root.seekg( 0, ios_base::beg );
        root_archive.resize(length);
        in_root.read((char*)root_archive.data(),length);
        in_root.close();
        
        ifstream in_boot(argv[2], ios::binary);
        if (!in_boot.is_open()) {
            glog.log("Could not open boot archive: %s\n",argv[2]);
            return 0;
        }
        in_boot.ignore( numeric_limits<streamsize>::max() );
        length = in_boot.gcount();
        in_boot.clear();
        in_boot.seekg( 0, ios_base::beg );
        boot_archive.resize(length);
        in_boot.read((char*)boot_archive.data(),length);
        in_boot.close();
        
        wib::Update req;
        glog.log("Sending root archive (%0.1f MB) and boot archive (%0.1f MB)\n",root_archive.size()/1024.0/1024.0,boot_archive.size()/1024.0/1024.0);
        req.set_root_archive(root_archive);
        req.set_boot_archive(boot_archive);
        wib::Empty rep;
        send_command(s,req,rep);
    } else if (cmd == "reboot") {
        wib::Reboot req;
        wib::Empty rep;
        send_command(s,req,rep);
    } else if (cmd == "peek") {
        if (argc != 2) {
            glog.log("Usage: peek addr\n");
            glog.log("   all arguments are base 16\n");
            return 0;
        }
        wib::Peek req;
        req.set_addr((size_t)strtoull(argv[1],NULL,16));
        wib::RegValue rep;
        send_command(s,req,rep);
        glog.log("0x%X\n",rep.value());
    } else if (cmd == "poke") {
        if (argc != 3) {
            glog.log("Usage: poke addr value\n");
            glog.log("   all arguments are base 16\n");
            return 0;
        }
        wib::Poke req;
        req.set_addr((size_t)strtoull(argv[1],NULL,16));
        req.set_value((uint32_t)strtoull(argv[2],NULL,16));
        wib::RegValue rep;
        send_command(s,req,rep);
    } else if (cmd == "cdpeek") {
        if (argc != 6) {
            glog.log("Usage: cdpeek femb_idx cd_idx chip_addr reg_page reg_addr\n");
            glog.log("   femb_idx and cd_idx are base 10, remaining args are base 16\n");
            return 0;
        }
        wib::CDPeek req;
        req.set_femb_idx((uint8_t)strtoull(argv[1],NULL,10));
        req.set_coldata_idx((uint8_t)strtoull(argv[2],NULL,10));
        req.set_chip_addr((uint8_t)strtoull(argv[3],NULL,16));
        req.set_reg_page((uint8_t)strtoull(argv[4],NULL,16));
        req.set_reg_addr((uint8_t)strtoull(argv[5],NULL,16));
        wib::CDRegValue rep;
        send_command(s,req,rep);
        glog.log("0x%X\n",rep.data());
    } else if (cmd == "cdpoke") {
        if (argc != 7) {
            glog.log("Usage: cdpoke femb_idx cd_idx chip_addr reg_page reg_addr data\n");
            glog.log("   femb_idx and cd_idx are base 10, remaining args are base 16\n");
            return 0;
        }
        wib::CDPoke req;
        req.set_femb_idx((uint8_t)strtoull(argv[1],NULL,10));
        req.set_coldata_idx((uint8_t)strtoull(argv[2],NULL,10));
        req.set_chip_addr((uint8_t)strtoull(argv[3],NULL,16));
        req.set_reg_page((uint8_t)strtoull(argv[4],NULL,16));
        req.set_reg_addr((uint8_t)strtoull(argv[5],NULL,16));
        req.set_data((uint8_t)strtoull(argv[6],NULL,16));
        wib::CDRegValue rep;
        send_command(s,req,rep);
    } else if (cmd == "cdfastcmd") {
        if (argc != 2) {
            glog.log("Usage: cdfastcmd cmd\n");
            glog.log("   cmd can be: reset, act, sync, edge, idle, edge_act\n");
            return 0;
        }
        wib::CDFastCmd req;
        string cmd(argv[1]);
        if (cmd == "reset") {
            req.set_cmd(FAST_CMD_RESET);
        } else if (cmd == "act") {
            req.set_cmd(FAST_CMD_ACT);
        } else if (cmd == "sync") {
            req.set_cmd(FAST_CMD_SYNC);
        } else if (cmd == "edge") {
            req.set_cmd(FAST_CMD_EDGE);
        } else if (cmd == "idle") {
            req.set_cmd(FAST_CMD_IDLE);
        } else if (cmd == "edge_act") {
            req.set_cmd(FAST_CMD_EDGE_ACT);
        } else {
            glog.log("Unknown fast command: %s\n",argv[1]);
            glog.log("Valid fast commands: reset, act, sync, edge, idle, edge_act\n");
            return 0;
        }
        wib::Empty rep;
        send_command(s,req,rep);
    } else if (cmd == "help") {
        print_help();
    } else {
        glog.log("Unrecognized Command: %s\n",argv[0]);
        return 0;
    }
    return 0;
}

int main(int argc, char **argv) {
    
    char *ip = (char*)"127.0.0.1";
    
    signed char opt;
    while ((opt = getopt(argc, argv, "w:h")) != -1) {
       switch (opt) {
           case 'h':
               print_usage(argv[0]);
               print_help();
               return 1;
           case 'w':
               ip = optarg;
               break;
           default: /* '?' */
               print_usage(argv[0]);
               return 1;
       }
    }
    
    zmq::context_t context(1);
    zmq::socket_t socket(context, ZMQ_REQ);
    
    char *addr;
    int len = asprintf(&addr,"tcp://%s:1234",ip);
    if (len < 0) return 1;
    
    socket.connect(addr);
    free(addr);
    
    if (optind < argc) {
        return run_command(socket,argc-optind,argv+optind);
    } else {
        char* buf;
        while ((buf = readline(">> ")) != nullptr) {
            if (strlen(buf) > 0) {
                add_history(buf);
            } else {
                free(buf);
                continue;
            }
            char *delim = (char*)" ";   
            int count = 1;
            char *ptr = buf;
            while((ptr = strchr(ptr, delim[0])) != NULL) {
                count++;
                ptr++;
            }
            if (count > 0) {
                char **cmd = new char*[count];
                cmd[0] = strtok(buf, delim);
                int i;
                for (i = 1; cmd[i-1] != NULL && i < count; i++) {
                    cmd[i] = strtok(NULL, delim);
                }
                if (cmd[i-1] == NULL) i--;
                int ret = run_command(socket,i,cmd);
                delete [] cmd;
                if (ret == 255) return 0;
                if (ret != 0) return ret;
            } else {
                return 0;
            }
            free(buf);
        }
    }
    
    return 0;
}

File: sw/wib_client.py
#!/usr/bin/env python3

import os
import sys
import time
import shlex
import pickle
import readline
import argparse

from wib import WIB
import wib_pb2 as wibpb

parser = argparse.ArgumentParser(description='Python version of the wib_client utility')
parser.add_argument('-w','--wib_server',default='127.0.0.1',help='IP of wib_server to connect to [127.0.0.1]')
parser.add_argument('command',nargs='?',help='Execute a single command (opens CLI if omitted; help command for more info)')
parser.add_argument('args',nargs='*',help='Optional arguments to command')

commands = argparse.ArgumentParser(add_help=False,prog='')
class CommandException(Exception):
    def __init__(self,message,parser=None):
        self.message = message
        self.parser = parser
def custom_exit(*args, message=None, **kwargs):
    raise CommandException(message)
def custom_error(message, *args, **kwargs):
    raise CommandException(message)
commands.exit = custom_exit
commands.error = custom_error
sub = commands.add_subparsers(title='subcommands',help='subcommand help')

def bind_parser(parser,func):
    def custom_exit(*args, message=None, **kwargs):
        raise CommandException(message,parser=parser)
    def custom_error(message, *args, **kwargs):
        raise CommandException(message,parser=parser)
    parser.exit = custom_exit
    parser.error = custom_error
    parser.set_defaults(func=func)

reboot_parser = sub.add_parser('reboot',help='Reboot the WIB',add_help=False)
def reboot(args):
    req = wibpb.Reboot()
    rep = wibpb.Empty()
    wib.send_command(req,rep)
    print('Rebooting...')
bind_parser(reboot_parser,reboot)

log_parser = sub.add_parser('log',help='Return or control the wib_server log',add_help=False)
log_parser.add_argument('action',nargs='?',choices=['clear','boot'],help='Optional log action (returns log if omitted)')
def log(args):
    req = wibpb.LogControl()
    req.return_log = args.action == None
    req.boot_log = args.action == 'boot'
    req.clear_log = args.action == 'clear'
    rep = wibpb.LogControl.Log()
    wib.send_command(req,rep)
    print(rep.contents.decode('ascii'),end='')
bind_parser(log_parser,log)

fw_timestamp_parser = sub.add_parser('fw_timestamp',help='Return firmware version timestamp',add_help=False)
def fw_timestamp(args):
    req = wibpb.GetTimestamp()
    rep = wibpb.GetTimestamp.Timestamp()
    wib.send_command(req,rep)
    print('fw_timestamp code: 0x%08X'%rep.timestamp);
    print('decoded: %i/%i/%i %i:%i:%i'%(rep.year,rep.month,rep.day,rep.hour,rep.min,rep.sec));
bind_parser(fw_timestamp_parser,fw_timestamp)

sw_version_parser = sub.add_parser('sw_version',help='Return software build version',add_help=False)
def sw_version(args):
    req = wibpb.GetSWVersion()
    rep = wibpb.GetSWVersion.Version()
    wib.send_command(req,rep)
    print('sw_version: %s'%rep.version);
bind_parser(sw_version_parser,sw_version)

timing_reset_parser = sub.add_parser('timing_reset',help='Reset the timing endpoint',add_help=False)
def timing_reset(args):
    req = wibpb.ResetTiming()
    rep = wibpb.GetTimingStatus.TimingStatus()
    wib.send_command(req,rep)
    wib.print_timing_status(rep)
bind_parser(timing_reset_parser,timing_reset)

timing_status_parser = sub.add_parser('timing_status',help='Return the status of the timing endpoint',add_help=False)
def timing_status(args):
    req = wibpb.GetTimingStatus()
    rep = wibpb.GetTimingStatus.TimingStatus()
    wib.send_command(req,rep)
    wib.print_timing_status(rep)
bind_parser(timing_status_parser,timing_status)

script_parser = sub.add_parser('script',help='Run a WIB script',add_help=False)
script_parser.add_argument('filename',help='local file will be sent, otherwise filename is remote in /etc/wib/ on the WIB')
def script(args):
    req = wibpb.Script()
    rep = wibpb.Status()
    if os.path.exists(args.filename):
        print('Executing local script...')
        with open(args.filename,'rb') as fin:
            req.script = fin.read()
        req.file = False
    else:
        print('Executing remote script...')
        req.script = bytes(args.filename, 'utf-8')
        req.file = True
    wib.send_command(req,rep)
    print('Successful:',rep.success)
bind_parser(script_parser,script)

config_parser = sub.add_parser('config',help='Send frontend configuration to the WIB',add_help=False)
config_parser.add_argument('filename',help='JSON config file')
def config(args):
    wib.config(args.filename)
bind_parser(config_parser,script)

log_parser = sub.add_parser('calibrate',help='Run the ADC calibration routine',add_help=False)
def log(args):
    req = wibpb.Calibrate()
    rep = wibpb.Status()
    wib.send_command(req,rep)
    print(rep.extra.decode('ascii'),end='')
    print('Successful:',rep.success)
bind_parser(log_parser,log)

daqspy_parser = sub.add_parser('daqspy',help='Read 1MB from each daq spy buffer and write the (up to) 2MB binary data',add_help=False)
daqspy_parser.add_argument('filename',help='Output file for binary data')
daqspy_parser.add_argument('buffers',nargs='?',choices=['buf0','buf1','both'],default='both',help='Select specific buffers [both]')
daqspy_parser.add_argument('--cmd',type=int,default=0,help='TLU commmand to use as trigge, or 0 for software trigger [0]')
daqspy_parser.add_argument('--rec',type=int,default=180360,help='Record time after trigger in 4.158ns ticks [180360 ~750us]')
daqspy_parser.add_argument('--timeout',type=int,default=60000,help='Trigger timeout time in ms [60000 ~1min]')
def daqspy(args):
    req = wibpb.ReadDaqSpy()
    req.buf0 = args.buffers in ['buf0','both']
    req.buf1 = args.buffers in ['buf1','both']
    req.trigger_command = args.cmd
    req.trigger_rec_ticks = args.rec
    req.trigger_timeout_ms = args.timeout
    rep = wibpb.ReadDaqSpy.DaqSpy()
    wib.send_command(req,rep)
    print('Successful:',rep.success)
    with open(args.filename,'wb') as fout:
        fout.write(rep.buf0)
        fout.write(rep.buf1)
bind_parser(daqspy_parser,daqspy)

peek_parser = sub.add_parser('peek',help='Read a 32bit value from WIB address space',add_help=False)
peek_parser.add_argument('addr',type=lambda x: int(x,16),help='Address to read (hex)')
def peek(args):
    req = wibpb.Peek()
    rep = wibpb.RegValue()
    req.addr = args.addr
    wib.send_command(req,rep)
    print('*(0x%016X) -> 0x%08X'%(rep.addr,rep.value))
bind_parser(peek_parser,peek)

poke_parser = sub.add_parser('poke',help='Write a 32bit value to WIB address space',add_help=False)
poke_parser.add_argument('addr',type=lambda x: int(x,16),help='Address to write (hex)')
poke_parser.add_argument('value',type=lambda x: int(x,16),help='Value to write to addr (hex)')
def poke(args):
    req = wibpb.Poke()
    rep = wibpb.RegValue()
    req.addr = args.addr
    req.value = args.value
    wib.send_command(req,rep)
    print('*(0x%016X) <- 0x%08X'%(rep.addr,rep.value))
bind_parser(poke_parser,poke)

cdpeek_parser = sub.add_parser('cdpeek',help='Read a 8bit value from COLDATA I2C address space',add_help=False)
cdpeek_parser.add_argument('femb_idx',type=int,choices=[0,1,2,3],help='FEMB to communicate with [0-3]')
cdpeek_parser.add_argument('coldata_idx',type=int,choices=[0,1],help='COLDDATA chip to communicate with [0-1]')
cdpeek_parser.add_argument('chip_addr',type=lambda x: int(x,16),help='DUNE I2C chip address (hex)')
cdpeek_parser.add_argument('reg_page',type=lambda x: int(x,16),help='DUNE I2C register page (hex)')
cdpeek_parser.add_argument('reg_addr',type=lambda x: int(x,16),help='DUNE I2C register address (hex)')
def cdpeek(args):
    req = wibpb.CDPeek()
    rep = wibpb.CDRegValue()
    req.femb_idx = args.femb_idx
    req.coldata_idx = args.coldata_idx
    req.chip_addr = args.chip_addr
    req.reg_page = args.reg_page
    req.reg_addr = args.reg_addr
    wib.send_command(req,rep)
    print('femb:%i coldata:%i chip:0x%02X page:0x%02X reg:0x%02X -> 0x%02X'%(rep.femb_idx,rep.coldata_idx,rep.chip_addr,rep.reg_page,rep.reg_addr,rep.data))
bind_parser(cdpeek_parser,cdpeek)

cdpoke_parser = sub.add_parser('cdpoke',help='Write a 8bit value to COLDATA I2C address space',add_help=False)
cdpoke_parser.add_argument('femb_idx',type=int,choices=[0,1,2,3],help='FEMB to communicate with')
cdpoke_parser.add_argument('coldata_idx',type=int,choices=[0,1],help='COLDDATA chip to communicate with')
cdpoke_parser.add_argument('chip_addr',type=int,help='DUNE I2C chip address')
cdpoke_parser.add_argument('reg_page',type=int,help='DUNE I2C register page')
cdpoke_parser.add_argument('reg_addr',type=int,help='DUNE I2C register address')
cdpoke_parser.add_argument('data',type=int,help='Data to write over DUNE I2C')
def cdpoke(args):
    req = wibpb.CDPoke()
    rep = wibpb.CDRegValue()
    req.femb_idx = args.femb_idx
    req.coldata_idx = args.coldata_idx
    req.chip_addr = args.chip_addr
    req.reg_page = args.reg_page
    req.reg_addr = args.reg_addr
    req.data = args.data
    wib.send_command(req,rep)
    print('femb:%i coldata:%i chip:0x%02X page:0x%02X reg:0x%02X <- 0x%02X'%(rep.femb_idx,rep.coldata_idx,rep.chip_addr,rep.reg_page,rep.reg_addr,rep.data))
bind_parser(cdpoke_parser,cdpoke)

cdfastcmd_parser = sub.add_parser('cdfastcmd',help='Send the fast command cmd to all coldata chips',add_help=False)
cdfastcmd_parser.add_argument('command',choices=['reset', 'act', 'sync', 'edge', 'idle', 'edge_act'],help='FASTCMD to send')
def cdfastcmd(args):
    fast_cmds = { 'reset':1, 'act':2, 'sync':4, 'edge':8, 'idle':16, 'edge_act':32 }
    req = wibpb.CDFastCmd()
    req.cmd = fast_cmds[args.command]
    rep = wibpb.Empty()
    wib.send_command(req,rep)
    print('Fast command sent')
bind_parser(cdfastcmd_parser,cdfastcmd)

update_parser = sub.add_parser('update',help='Deploy a new root and boot archive to the WIB',add_help=False)
update_parser.add_argument('root_archive',help='Root filesystem archive')
update_parser.add_argument('boot_archive',help='Boot filesystem archive')
def update(args):
    if not os.path.exists(args.root_archive):
        print('Root archive',args.root_archive,'not found. Aborting update.')
        return
    if not os.path.exists(args.boot_archive):
        print('Boot archive',args.boot_archive,'not found. Aborting update.')
        return
    req = wibpb.Update()
    with open(args.root_archive,'rb') as froot:
        req.root_archive = froot.read()
    with open(args.boot_archive,'rb') as fboot:
        req.boot_archive = fboot.read()
    rep = wibpb.Empty()
    print('Sending update command...')
    wib.send_command(req,rep) 
    print('WIB will now update and reboot.')
bind_parser(update_parser,update)

exit_parser = sub.add_parser('exit',help='Closes the command interface',add_help=False)
def exit(args):
    sys.exit(0)
bind_parser(exit_parser,exit)

help_parser = sub.add_parser('help',help='Show help message, optionally for a subcommand',add_help=False)
help_parser.add_argument('command',nargs='?',help='Subcommand to show help for')
def help(args):
    if args.command is None:
        commands.print_help()
    else:
        p = '%s_parser'%args.command
        if p in globals():
            eval('%s.print_help()'%p)
        else:
            print('Unknown command %s'%args.command)
bind_parser(help_parser,help)

def handle_args(args):
    try:
        args = commands.parse_args(args)
        args.func(args)
    except CommandException as ex:
        if ex.parser is not None:
            ex.parser.print_usage()
        print(ex.message)

if __name__ == "__main__":
    args = parser.parse_args()
    wib = WIB(args.wib_server)
    if args.command is None:
        import readline
        hist_file = os.path.expanduser('~/.wib_client_hist')
        if os.path.exists(hist_file):
            readline.read_history_file(hist_file)
        else:
            open(hist_file,'w').close()
        hist_start = readline.get_current_history_length()
        while True:
            try:
                line = input('[%s] >> '%args.wib_server).strip()
                handle_args(shlex.split(line))
            except KeyboardInterrupt:
                print()
                continue
            except EOFError:
                print()
                break
        hist_end = readline.get_current_history_length()
        readline.append_history_file(hist_end-hist_start,hist_file)
    else:
        handle_args([args.command]+args.args)
